src/xrpld/app/tx/detail/VaultCreate.cpp

```diff
@@ -0,0 +1,198 @@
+            return temMALFORMED;  // DomainID only allowed on private vaults
+    }
+
+    if (auto const metadata = ctx.tx[~sfMPTokenMetadata])
+    {
+        if (metadata->length() == 0 ||
+            metadata->length() > maxMPTokenMetadataLength)
+            return temMALFORMED;
+    }
+
+    return preflight2(ctx);
+}
+
+XRPAmount
+VaultCreate::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    // One reserve increment is typically much greater than one base fee.
+    return view.fees().increment;
+}
+
+TER
+VaultCreate::preclaim(PreclaimContext const& ctx)
+{
+    auto asset = ctx.tx[sfAsset];
+    auto account = ctx.tx[sfAccount];
+
+    if (asset.holds<MPTIssue>())
+    {
+        auto mptID = asset.get<MPTIssue>().getMptID();
+        auto issuance = ctx.view.read(keylet::mptIssuance(mptID));
+        if (!issuance)
+            return tecNO_ENTRY;
+        if (issuance->getFlags() & lsfMPTLocked)
+            return tecLOCKED;
+        if ((issuance->getFlags() & lsfMPTCanTransfer) == 0)
+            return tecNO_AUTH;
+    }
+
+    // Cannot create Vault for an Asset frozen for the vault owner
+    if (isFrozen(ctx.view, account, asset))
+        return tecFROZEN;
+
+    if (auto const domain = ctx.tx[~sfDomainID])
+    {
+        auto const sleDomain =
+            ctx.view.read(keylet::permissionedDomain(*domain));
+        if (!sleDomain)
+            return tecNO_ENTRY;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+VaultCreate::doApply()
+{
+    // All return codes in `doApply` must be `tec`, `ter`, or `tes`.
+    // As we move checks into `preflight` and `preclaim`,
+    // we can consider downgrading them to `tef` or `tem`.
+
+    auto const& tx = ctx_.tx;
+    auto const& ownerId = account_;
+    auto sequence = tx.getSeqValue();
+
+    auto owner = view().peek(keylet::account(ownerId));
+    auto vault = std::make_shared<SLE>(keylet::vault(ownerId, sequence));
+
+    if (auto ter = dirLink(view(), ownerId, vault))
+        return ter;
+    // Should the next 3 lines be folded into `dirLink`?
+    adjustOwnerCount(view(), owner, 1, j_);
+    auto ownerCount = owner->at(sfOwnerCount);
+    if (mPriorBalance < view().fees().accountReserve(ownerCount))
+        return tecINSUFFICIENT_RESERVE;
+
+    auto maybePseudo = createPseudoAccount(
+        view(), vault->key(), PseudoAccountOwnerType::Vault);
+    if (!maybePseudo)
+        return maybePseudo.error();
+    auto& pseudo = *maybePseudo;
+    auto pseudoId = pseudo->at(sfAccount);
+
+    if (auto ter =
+            addEmptyHolding(view(), pseudoId, mPriorBalance, tx[sfAsset], j_))
+        return ter;
+
+    auto txFlags = tx.getFlags();
+    std::uint32_t mptFlags = 0;
+    if ((txFlags & tfVaultShareNonTransferable) == 0)
+        mptFlags |= (lsfMPTCanEscrow | lsfMPTCanTrade | lsfMPTCanTransfer);
+    if (txFlags & tfVaultPrivate)
+        mptFlags |= lsfMPTRequireAuth;
+
+    auto maybeShare = MPTokenIssuanceCreate::create(
```

Comment:
Same here, I'm guessing this isn't finished though because this only handles MPT tokens.

Reply:
This MTP is unrelated to the assets held by the vault. This MPT is for shares in the vault.

The MPToken or TrustLine/RippleState for the assets held by the vault is created above, in `addEmptyHolding`